src/Artifacts.jl

```diff
@@ -0,0 +1,589 @@
+
+    # First, extract OS; we need to build a mapping here
+    os_map = Dict(
+        "windows" => Windows,
+        "macos" => MacOS,
+        "freebsd" => FreeBSD,
+        "linux" => Linux,
+    )
+    P = get(os_map, lowercase(entry["os"]), UnknownPlatform)
+
+    # Next, architecture, libc, libgfortran version and cxxabi (if given)
+    arch = nosym(get(entry, "arch", nothing))
+    libc = nosym(get(entry, "libc", nothing))
+    libgfortran_version = nover(get(entry, "libgfortran_version", nothing))
+    libstdcxx_version = nover(get(entry, "libstdcxx_version", nothing))
+    cxxstring_abi = nosym(get(entry, "cxxstring_abi", nothing))
+
+    # Construct the actual Platform object
+    return P(arch;
+        libc=libc,
+        compiler_abi=CompilerABI(
+            libgfortran_version=libgfortran_version,
+            libstdcxx_version=libstdcxx_version,
+            cxxstring_abi=cxxstring_abi
+        ),
+    )
+end
+
+function pack_platform!(meta::Dict, p::Platform)
+    os_map = Dict(
+        Windows => "windows",
+        MacOS => "macos",
+        FreeBSD => "freebsd",
+        Linux => "linux",
+    )
+    meta["os"] = os_map[typeof(p)]
+    meta["arch"] = string(arch(p))
+    if libc(p) != nothing
+        meta["libc"] = string(libc(p))
+    end
+    if libgfortran_version(p) != nothing
+        meta["libgfortran_version"] = string(libgfortran_version(p))
+    end
+    if libstdcxx_version(p) != nothing
+        meta["libstdcxx_version"] = string(libstdcxx_version(p))
+    end
+    if cxxstring_abi(p) != nothing
+        meta["cxxstring_abi"] = string(cxxstring_abi(p))
+    end
+end
+
+"""
+    artifact_meta(name::String, artifact_toml::String;
+                  platform::Platform = platform_key_abi())
+
+Get metadata about a given artifact (identified by name) stored within the given
+`Artifact.toml` file.  If the artifact is platform-specific, use `platform` to choose the
+most appropriate mapping.  If none is found, return `nothing`.
+"""
+function artifact_meta(name::String, artifact_toml::String;
+                       platform::Platform = platform_key_abi())
+    if !isfile(artifact_toml)
+        return nothing
+    end
+
+    # Parse the toml for the 
+    return artifact_meta(name, parse_toml(artifact_toml), artifact_toml; platform=platform)
+end
+
+function artifact_meta(name::String, artifact_dict::Dict, artifact_toml::String;
+                       platform::Platform = platform_key_abi())
+    if !haskey(artifact_dict, name)
+        return nothing
+    end
+    meta = artifact_dict[name]
+
+    # If it's an array, find the entry that best matches our current platform
+    if isa(meta, Array)
+        dl_dict = Dict(unpack_platform(x, name, artifact_toml) => x for x in meta)
+        meta = select_platform(dl_dict, platform)
+    
+    # If it's NOT a dict, complain
+    elseif !isa(meta, Dict)
+        @warn("Invalid Artifact.toml at $(artifact_toml): artifact '$name' malformed, must be array or dict!")
```

Comment:
Should these `@warn` be `@error`?

Reply:
Since `artifact_meta()` already has a nice failure condition (return `nothing`) I figured it was better to allow programs to continue, assuming that the artifact does not exist if something is horribly wrong with their `Artifact.toml` instead of just error'ing out.  I'm not married to that design decision though.

Reply:
`@error` is not the same as `error`. It is just whether to print in red or yellow and call it "Warning" vs "Error".

<img width="211" alt="image" src="https://user-images.githubusercontent.com/1282691/63109766-39f10680-bf8a-11e9-8e99-37fe18ae72c1.png">

Reply:
Oh!  I see what you mean.  Yes, I do think that's better.  Red is better here.